Connection read timeout hint

fbiville opened the pull request:
Companion PR: https://github.com/neo4j-drivers/testkit/pull/233

fbiville committed:
Implement connection read timeout hint

Reset connection read deadline after every message reception,
including NOOPs.

bigmontz commented on neo4j/internal/bolt/bolt4.go:
```diff
@@ -182,8 +184,9 @@ func (b *bolt4) receiveMsg() interface{} {
-	msg, err := b.in.next(b.conn)
+	msg, err := b.in.next(b.conn, b.resetConnectionReadDeadline)
 	b.setError(err, true)
+	b.resetConnectionReadDeadline()
```
Seems like we are sharing the responsibility of calling `resetConnectionReadDeadline` between `Bolt` and `dechunker`.  This responsibility should be or pushed to the `dechunker` or keep it in bolt. I tend to prefer to push it to the lower level (`dechunker`) since this one is handling the actual reading.

fbiville replied:
yep, good idea... this first draft felt messy.
The only drawback I see is that the dechunker will now have to accept a `net.Conn` instead of the currently more generic `io.Reader`

fbiville replied:
I pushed a new version. Migrating tests from `io.Reader` to `net.Conn` was a bit of a PITA, but I finally managed to do it.

fbiville committed:
Move connection deadline reset to dechunker only

Adapt unit tests, since the dechunker now deals with net.Conn, not
io.Reader. Tests now rely on net.Pipe()

bigmontz commented on neo4j/internal/bolt/dechunker_test.go:
```diff
@@ -69,19 +70,26 @@ func TestDechunker(t *testing.T) {
 		for i := range msgBuf {
 			if msgBuf[i] != b {
 				t.Errorf("Wrong content in buffer at %d, %d vs %d (message %d)", i, msgBuf[i], b, msgi)
+				AssertNoError(t, serv.Close())
```
Is that a way to unit test the `dechunker` reset timeout logic?

fbiville replied:
I only relied on the existing Testkit tests. Is it worth adding a test here as well in your opinion?

fbiville replied:
And now there is an integration test as well

bigmontz commented on neo4j/internal/bolt/hydrator.go:
```diff
@@ -241,6 +242,9 @@ func (h *hydrator) success(n uint32) *success {
 			succ.notifications = parseNotifications(l)
 		case "rt":
 			succ.routingTable = h.routingTable()
+		case "hints":
```
Is there test for the hint set?

fbiville replied:
Via testkit, yes.

fbiville replied:
And now a unit test as well

bigmontz reviewed (approved):
🧀

fbiville committed:
Test timeouts

fbiville committed:
Test configuration hint parsing

bigmontz reviewed (approved):
📦

fbiville committed:
Exclude routing-aware configuration hint testkit tests

`GetRoutingTable` is too complicated to implement in Go
